lumicks/pylake/kymotracker/detail/velocity.py

```diff
@@ -0,0 +1,254 @@
+        Maximum number of iterations allowed in the optimization algorithm. Ignored if
+        `n_breakpoints == 0`.
+    n_restarts : int
+        Number of times to run the optimization algorithm with breakpoint initial guesses drawn
+        from a uniform distribution. Ignored if `n_breakpoints == 0`.
+
+    Returns
+    -------
+    PiecewiseModel
+        Parameters and optimization statistics for piecewise linear fit.
+
+    References
+    ----------
+    .. [1] Muggeo, V. (2003) Estimating Regression Models with Unknown Break-Points. Statistics in
+           Medicine, 22(19), 3055-3071.
+    """
+
+    if n_breakpoints < 0:
+        raise ValueError(f"Number of requested breakpoints must be >= 0, got {n_breakpoints}.")
+    elif n_breakpoints == 0:
+        return _optimize_linear(x, y)
+
+    results = [_optimize_breakpoints(x, y, n_breakpoints, tol, max_iter) for _ in range(n_restarts)]
+    filtered_results = [result for result in results if result.converged]
+
+    # todo: raise if model out of bounds
+    # find best converged fit, fallback to un-converged if necessary
+    return min(filtered_results if len(filtered_results) else results, key=lambda r: r.rss)
+
+
+def _optimize_linear(x, y):
+    """OLS linear regression of x and y.
+
+    Parameters
+    ----------
+    x : np.ndarray
+        Dependent variable (time).
+    y : np.ndarray
+        Independent variable (position).
+    """
+    n_coeffs = 2
+    n_samples = len(x)
+
+    design_matrix = np.vstack((np.ones(n_samples), x)).T
+    xtx = np.linalg.pinv(np.matmul(design_matrix.T, design_matrix))
+    coeffs = np.dot(
+        xtx,
+        np.dot(design_matrix.T, y),
+    )
+
+    fit = np.dot(design_matrix, coeffs)
+    residuals = fit - y
+    rss = np.sum(residuals**2)
+    bic = n_samples * np.log(rss / n_samples) + n_coeffs * np.log(n_samples)
+
+    intercept, slope = coeffs
+
+    cov = (rss / (n_samples - n_coeffs)) * xtx
+    intercept_std, slope_std = np.sqrt(np.diag(cov))
+
+    return PiecewiseModel(
+        intercept, [slope], [], intercept_std, [slope_std], [], rss, bic, "converged"
+    )
+
+
+def _optimize_breakpoints(x, y, n_breakpoints, tol, max_iter):
+    r"""
+    Optimization of breakpoint locations for a continuous piecewise linear model
+
+    Parameters
+    ----------
+    x : np.ndarray
+        Dependent variable (time).
+    y : np.ndarray
+        Independent variable (position).
+    n_breakpoints : int
+        Number of requested breakpoints.
+    tol : float
+        Convergence tolerance. Ignored if `n_breakpoints == 0`.
+    max_iter : int
+        Maximum number of iterations allowed in the optimization algorithm. Ignored if
+        `n_breakpoints == 0`.
+    n_restarts : int
+        Number of times to run the optimization algorithm with breakpoint initial guesses drawn
+        from a uniform distribution. Ignored if `n_breakpoints == 0`.
+    """
+    n_samples = len(x)
+    n_coeffs = 2 + 2 * n_breakpoints
```

Comment:
I do wonder if BIC treating the breakpoint as just another parameter is sufficient penalty. In a sense, it is a very different type of parameter. See also the answer on this for example: https://stats.stackexchange.com/questions/337852/aic-bic-for-a-segmented-regression-model

When you run this with many simulations, how consistently does the BIC selection procedure recover the correct model?